### PR TITLE
Update sam_expect to be used for resource attributes

### PR DESCRIPTION
**Pull request by aaythapa**
### Issue #, if available

Add ability to use `sam_expect` on Resource attributes like Properties. Going to be used on the `Connectors` attribute too. Still needs some refactoring for it be used on things like DeletionPolicy and Type.

### Checklist

- [x] Adheres to the [development guidelines](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#development-guidelines)
- [ ] Add/update [transform tests](https://github.com/aws/serverless-application-model/blob/develop/DEVELOPMENT_GUIDE.md#unit-testing-with-multiple-python-versions)
    - [ ] Using correct values
    - [ ] Using wrong values
- [ ] Add/update [integration tests](https://github.com/aws/serverless-application-model/blob/develop/INTEGRATION_TESTS.md)

### Examples?

Please reach out in the comments if you want to add an example. Examples will be 
added to `sam init` through [aws/aws-sam-cli-app-templates](https://github.com/aws/aws-sam-cli-app-templates).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: samtranslator/model/exceptions.py [hunk 1]
```diff
@@ -102,21 +102,40 @@ class InvalidResourcePropertyTypeException(InvalidResourceException):
...
+    ) -> None:
+        message = message or self._default_message(logical_id, key_path, expected_type)
+        super().__init__(logical_id, message)
+
+    @staticmethod
+    def _default_message(logical_id: str, key_path: str, expected_type: Optional[ExpectedType]) -> str:
+        if expected_type:
+            type_description, _ = expected_type.value
+            return f"Attribute '{key_path}' for resource '{logical_id}' should be a {type_description}."
+        return f"Type of attribute '{key_path}' for resource '{logical_id}'is invalid."
```
Comment (hoffa) [inline samtranslator/model/exceptions.py:138]:
```suggestion
        return f"Type of attribute '{key_path}' for resource '{logical_id}' is invalid."
```

Comment (hoffa) [reply samtranslator/model/exceptions.py:138]: Where will this be used?

Comment (aaythapa) [reply samtranslator/model/exceptions.py:138]: For the `Connectors` attribute and for other attributes like `Type` and `DeletionPolicy`. Those other attributes are in the `resource.py` file which needs some refactoring for it to use `sam_expect` (I left a TODO there). Going to open another PR for that and do it later on.

File: samtranslator/parser/parser.py [hunk 2]
```diff
@@ -41,16 +46,15 @@ def validate_datatypes(sam_template):  # type: ignore[no-untyped-def]
...
-                            "All 'Resources' must be Objects and have a 'Properties' Object. If "
-                            "you're using YAML, this may be an indentation issue.",
-                        )
-                    ]
+            try:
+                sam_expect(
+                    sam_resource.properties, resource_logical_id, "Properties", is_resource_attribute=True
+                ).to_be_a_map(
+                    "All 'Resources' must be Objects and have a 'Properties' Object. If "
+                    "you're using YAML, this may be an indentation issue."
```
Comment (aahung) [inline samtranslator/parser/parser.py:54]: This is too verbose IMO. Can we remove this custom message and leave it default?

File: samtranslator/model/exceptions.py [hunk 3]
```diff
@@ -102,21 +102,40 @@ class InvalidResourcePropertyTypeException(InvalidResourceException):
...
+    ) -> None:
+        message = message or self._default_message(logical_id, key_path, expected_type)
+        super().__init__(logical_id, message)
+
+    @staticmethod
+    def _default_message(logical_id: str, key_path: str, expected_type: Optional[ExpectedType]) -> str:
+        if expected_type:
+            type_description, _ = expected_type.value
+            return f"Attribute '{key_path}' for resource '{logical_id}' should be a {type_description}."
+        return f"Type of attribute '{key_path}' for resource '{logical_id}' is invalid."
```
Comment (hoffa) [inline samtranslator/model/exceptions.py:138]: Do we have to mention logical ID? In the test example the logical ID is already specified ("Resource with id [Function] is invalid.").

Comment (aaythapa) [reply samtranslator/model/exceptions.py:138]: You're right, its not mentioned when resource property type is invalid so it shouldn't be mentioned here.

**Review by hoffa — APPROVED** [review]

**Review by aahung — APPROVED** [review]

**Pull request merged**